File: golang/121. Best Time to Buy and Sell Stock .go
func maxProfit(prices []int) int {
    if len(prices) == 0 {
        return 0
    }
    
    max, level := 0, prices[0]
    for _, price := range prices[1:] { 
        max = maximum(max, price - level)
        if price < level {
            level = price
        }
    }
    
    return max 
}

func maximum(a int, b int) int {
    if a > b {
        return a
    }
    return b
}

File: java/121. Best Time to Buy and Sell Stock .java
class Solution {
    public int maxProfit(int[] prices) {
        
        int max = Integer.MIN_VALUE;
        int level = 0;
        
        for(int i = 0; i < prices.length - 1; i++){
            if(level < 0) level = 0;
            level += prices[i + 1] - prices[i];
            max = Math.max(level, max);
        }
        
        return max < 0 ? 0 : max;
    }
}
